File: api/src/main/java/fr/huiitre/tools/modules/riot/valorant/application/core/usecase/UnlinkValorantAccountUseCase.java
package fr.huiitre.tools.modules.riot.valorant.application.core.usecase;

import fr.huiitre.tools.modules.core.module.domain.ModuleCode;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.core.security.application.ports.AuthenticatedUserProvider;
import fr.huiitre.tools.modules.core.security.application.usecase.SecuredUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.core.ports.ValorantAuthRepository;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
public class UnlinkValorantAccountUseCase implements SecuredUseCase {

    private final ValorantAuthRepository valorantAuthRepository;
    private final AuthenticatedUserProvider authenticatedUserProvider;

    public UnlinkValorantAccountUseCase(ValorantAuthRepository valorantAuthRepository,
                                         AuthenticatedUserProvider authenticatedUserProvider) {
        this.valorantAuthRepository = valorantAuthRepository;
        this.authenticatedUserProvider = authenticatedUserProvider;
    }

    @Override
    public Optional<ModuleCode> requiredModule() {
        return Optional.of(ModuleCode.RIOT);
    }

    @Override
    public RoleCode requiredRole() {
        return RoleCode.USER;
    }

    public void execute(Long accountId) {
        Long userId = authenticatedUserProvider.getUserId();
        if (!valorantAuthRepository.existsByIdAndUserId(accountId, userId)) {
            throw new IllegalArgumentException("VALORANT_ACCOUNT_NOT_FOUND");
        }
        valorantAuthRepository.deleteById(accountId);
    }
}

File: api/Modules/Riot/Valorant/Application/Core/Usecases/UnlinkValorantAccountUseCase.cs
using Tools.Api.Modules.Core.Common.Application.Exceptions;
using Tools.Api.Modules.Core.Security.Application.Services;
using Tools.Api.Modules.Core.Security.Application.Usecases;
using Tools.Api.Modules.Core.Security.Domain;
using Tools.Api.Modules.Riot.Valorant.Application.Core.Ports;

namespace Tools.Api.Modules.Riot.Valorant.Application.Core.Usecases;

public sealed class UnlinkValorantAccountUseCase(
    UseCaseAuthorizer authorizer,
    IValorantAuthRepository valorantAuthRepository
) : SecuredUseCase(authorizer)
{
    protected override RoleCode RequiredRole => RoleCode.User;
    protected override ModuleCode? RequiredModule => ModuleCode.Riot;

    // La suppression emporte watchlist, skins possédés et historique : ils sont en cascade sur
    // valorant_account.
    public async Task Execute(long accountId)
    {
        if (!await valorantAuthRepository.ExistsByIdAndUserId(accountId, CurrentUser.UserId))
        {
            throw AppException.NotFound(
                "VALORANT_ACCOUNT_NOT_FOUND",
                "Ce compte Valorant est introuvable.");
        }

        await valorantAuthRepository.DeleteById(accountId);
    }
}
